core/src/main/java/com/linecorp/armeria/common/multipart/MultipartDecoder.java

```diff
@@ -98,16 +129,242 @@ public long demand() {
     @Override
     public void subscribe(Subscriber<? super BodyPart> subscriber, EventExecutor executor,
                           SubscriptionOption... options) {
-        decoded.subscribe(subscriber, executor, options);
+        requireNonNull(subscriber, "subscriber");
+        requireNonNull(executor, "executor");
+        requireNonNull(options, "options");
+
+        if (!subscribedUpdater.compareAndSet(this, 0, 1)) {
+            if (executor.inEventLoop()) {
+                abortLateSubscriber(subscriber, abortCause);
+            } else {
+                executor.execute(() -> abortLateSubscriber(subscriber, abortCause));
+            }
+            return;
+        }
+        final MultipartSubscriber delegatedSubscriber = new MultipartSubscriber(subscriber, executor);
+        this.delegatedSubscriber = delegatedSubscriber;
+        decoded.subscribe(delegatedSubscriber, executor, options);
+        subscribedCompleteFuture.complete(null);
+    }
+
+    private static void abortLateSubscriber(Subscriber<? super BodyPart> subscriber,
+                                            @Nullable Throwable throwable) {
+        subscriber.onSubscribe(NoopSubscription.get());
+        if (throwable != null) {
+            subscriber.onError(throwable);
+        } else {
+            subscriber.onError(new IllegalStateException("subscribed by other subscriber already"));
+        }
     }
 
     @Override
     public void abort() {
-        decoded.abort();
+        abort(AbortedStreamException.get());
     }
 
     @Override
     public void abort(Throwable cause) {
-        decoded.abort(cause);
+        requireNonNull(cause, "cause");
+        if (subscribed == -1) {
+            // Already aborted before
+            return;
+        }
+        abortCause = cause;
+        if (subscribedUpdater.compareAndSet(this, 0, -1)) {
+            // There is no subscriber
+            return;
+        }
+        subscribedCompleteFuture.handle((unused, throwable) -> {
+            decoded.abort(cause);
```

Comment:
Can't we just abort `decoded` without using the `subscribedCompleteFuture`?
The subscriber would get `AbortedStreamException` if `decoded.abort(cause);` happens before the subscription which is fine.